File: CSharp/Ops/OPSConfig.cs
///////////////////////////////////////////////////////////
//  OPSConfig.cs
//  Implementation of the Class OPSConfig
//  Created on:      12-nov-2011 09:25:33
//  Author:
///////////////////////////////////////////////////////////

using System.Collections.Generic;       // Needed for the "List"
using System.IO;
using System.Text;

namespace Ops 
{
	public class OPSConfig : OPSObject 
    {
        public const byte OPSConfig_idlVersion = 0;
        private byte _OPSConfig_version = OPSConfig_idlVersion;
        public byte OPSConfig_version
        {
            get { return _OPSConfig_version; }
            set {
                if (value > OPSConfig_idlVersion)
                {
                    throw new IdlVersionException(
                        "OPSConfig: received version '" + value + "' > known version '" + OPSConfig_idlVersion + "'");
                }
                _OPSConfig_version = value;
            }
        }

        private protected List<Domain> domains = new List<Domain>();

        public static OPSConfig GetConfig() 
        {
            return OPSConfigRepository.GetConfig(); //GetConfig("ops_config.xml");
        }

        public static OPSConfig GetConfig(string configFile)
        {
            try
            {
                FileStream fis = File.OpenRead(configFile);
                XMLArchiverIn archiverIn = new XMLArchiverIn(fis, "root");
                archiverIn.Add(OPSObjectFactory.GetInstance());
                OPSConfig newConfig = null;
                newConfig = (OPSConfig)archiverIn.Inout("ops_config", newConfig);
                fis.Close();
                return newConfig;
            }
            catch (FileNotFoundException)
            {
                Logger.ExceptionLogger.LogMessage("OPSConfig::GetConfig(), File NOT found: " + configFile);
                return null;
            }
        }

        public static void SaveConfig(OPSConfig conf, string configFile) 
        {
            FileStream fos = File.OpenWrite(configFile);
            BinaryWriter bw = new BinaryWriter(fos);
            XMLArchiverOut archiverOut = new XMLArchiverOut(bw, false);
            bw.Write(Encoding.ASCII.GetBytes("<root xmlns=ops>\n"));
            archiverOut.Inout("ops_config", conf);
            bw.Write(Encoding.ASCII.GetBytes("</root>"));
            fos.Close();
        }

        //---------------------------------------------------------------

        public OPSConfig()
        {
            AppendType("OPSConfig");
        }

        public Domain GetDomain(string domainID)
        {
            foreach (Domain domain in domains)
            {
                if (domain.GetDomainID().Equals(domainID))
                {
                    return domain;
                }
            }
            return null;
        }

        public override void Serialize(IArchiverInOut archive)
        {
            base.Serialize(archive);
            if (IdlVersionMask != 0)
            {
                OPSConfig_version = archive.Inout("OPSConfig_version", OPSConfig_version);
            }
            else
            {
                OPSConfig_version = 0;
            }
            domains = (List<Domain>)archive.InoutSerializableList("domains", domains);
        }

        public List<Domain> getDomains()
        {
            return domains;
        }

        public override object Clone()
        {
            OPSConfig cloneResult = new OPSConfig();
            FillClone(cloneResult);
            return cloneResult;
        }

        public override void FillClone(OPSObject cloneO)
        {
            base.FillClone(cloneO);
            OPSConfig cloneResult = (OPSConfig)cloneO;
            cloneResult.OPSConfig_version = OPSConfig_version;
            cloneResult.domains = new List<Domain>(domains.Count);
            domains.ForEach((item) => { cloneResult.domains.Add((Domain)item.Clone()); });
        }
    }

}

File: Java/OPSJLib/src/ops/OPSConfig.java
/**
*
* Copyright (C) 2006-2009 Anton Gravestam.
* Copyright (C) 2020-2024 Lennart Andersson.
*
* This file is part of OPS (Open Publish Subscribe).
*
* OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.

* OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
*/
package ops;

import configlib.ArchiverInOut;
import configlib.XMLArchiverIn;
import configlib.XMLArchiverOut;
import configlib.exception.FormatException;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Vector;
import ops.archiver.OPSObjectFactory;
import ops.OPSConfigRepository;

/**
 *
 * @author angr
 */
public class OPSConfig extends OPSObject
{
    public byte OPSConfig_version = OPSConfig_idlVersion;

    public static final byte OPSConfig_idlVersion = 0;

    //static OPSConfig theConfig = null;
    public static OPSConfig getConfig() throws IOException, FormatException
    {
        return OPSConfigRepository.getConfig(); // getConfig(new File("ops_config.xml"));
    }

    public static OPSConfig getConfig(File configFile) throws IOException, FormatException
    {
        FileInputStream fis = new FileInputStream(configFile);
        XMLArchiverIn archiverIn = new XMLArchiverIn(fis, "root");
        archiverIn.add(OPSObjectFactory.getInstance());
        OPSConfig newConfig = null;
        newConfig = (OPSConfig) archiverIn.inout("ops_config", newConfig);

        return newConfig;
    }

    public static void saveConfig(OPSConfig conf, File configFile) throws FileNotFoundException, IOException
    {
        FileOutputStream fos = new FileOutputStream(configFile);
        XMLArchiverOut archiverOut = new XMLArchiverOut(fos, false);
        fos.write("<root xmlns=ops>\n".getBytes());
            archiverOut.inout("ops_config", conf);
        fos.write("\n</root>".getBytes());
    }

    //------------------------------
    protected Vector<Domain> domains = new Vector<Domain>();

    public OPSConfig()
    {
        appendType("OPSConfig");
    }

    public Domain getDomain(String domainID)
    {
        for (Domain domain : domains)
        {
            if (domain.getDomainID().equals(domainID))
            {
                return domain;
            }
        }
        return null;
    }

    @Override
    public void serialize(ArchiverInOut archive) throws IOException
    {
        super.serialize(archive);
        if (idlVersionMask != 0) {
            byte tmp = archive.inout("OPSConfig_version", OPSConfig_version);
            if (tmp > OPSConfig_idlVersion) {
                throw new IOException("OPSConfig: received version '" + tmp + "' > known version '" + OPSConfig_idlVersion + "'");
            }
            OPSConfig_version = tmp;
        } else {
            OPSConfig_version = 0;
        }
        domains = (Vector<Domain>) archive.inoutSerializableList("domains", domains);
    }

    public Vector<Domain> getDomains()
    {
        return domains;
    }

    @Override
    public Object clone()
    {
        OPSConfig cloneResult = new OPSConfig();
        fillClone(cloneResult);
        return cloneResult;
    }

    @Override
    public void fillClone(OPSObject cloneO)
    {
        super.fillClone(cloneO);
        OPSConfig cloneResult = (OPSConfig)cloneO;
        cloneResult.OPSConfig_version = this.OPSConfig_version;
        cloneResult.domains = new java.util.Vector<Domain>();
        this.domains.forEach((item) -> cloneResult.domains.add((Domain)item.clone()));
    }

}
